core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDecider.java

```diff
@@ -162,21 +166,49 @@ public void onNewInfo(ClusterInfo info) {
             Map<String, DiskUsage> usages = info.getNodeDiskUsages();
             if (usages != null) {
                 boolean reroute = false;
-                for (DiskUsage entry : usages.values()) {
-                    warnAboutDiskIfNeeded(entry);
-                    if (entry.getFreeBytes() < DiskThresholdDecider.this.freeBytesThresholdHigh.bytes() ||
-                            entry.getFreeDiskAsPercentage() < DiskThresholdDecider.this.freeDiskThresholdHigh) {
+                String explanation = "";
+                for (Map.Entry<String, DiskUsage> entry : usages.entrySet()) {
+                    String node = entry.getKey();
+                    DiskUsage usage = entry.getValue();
+                    warnAboutDiskIfNeeded(usage);
+                    if (usage.getFreeBytes() < DiskThresholdDecider.this.freeBytesThresholdHigh.bytes() ||
+                            usage.getFreeDiskAsPercentage() < DiskThresholdDecider.this.freeDiskThresholdHigh) {
                         if ((System.nanoTime() - lastRunNS) > DiskThresholdDecider.this.rerouteInterval.nanos()) {
                             lastRunNS = System.nanoTime();
                             reroute = true;
+                            explanation = "high disk watermark exceeded on one or more nodes";
                         } else {
                             logger.debug("high disk watermark exceeded on {} but an automatic reroute has occurred in the last [{}], skipping reroute",
-                                    entry, DiskThresholdDecider.this.rerouteInterval);
+                                    node, DiskThresholdDecider.this.rerouteInterval);
+                        }
+                        nodeHasPassedHighWatermark.put(node, true);
+                    } else if (usage.getFreeBytes() < DiskThresholdDecider.this.freeBytesThresholdLow.bytes() ||
+                            usage.getFreeDiskAsPercentage() < DiskThresholdDecider.this.freeDiskThresholdLow) {
+                        nodeHasPassedLowWatermark.put(node, true);
+                    } else {
+                        Boolean passedHigh = nodeHasPassedHighWatermark.get(node);
+                        Boolean passedLow = nodeHasPassedLowWatermark.get(node);
+
+                        if ((passedHigh != null && passedHigh) || (passedLow != null && passedLow)) {
+                            // The entry has previously been over the high or
+                            // low watermark, but is no longer, so we should
+                            // reroute so any unassigned shards can be allocated
+                            // if they are able to be
+                            if ((System.nanoTime() - lastRunNS) > DiskThresholdDecider.this.rerouteInterval.nanos()) {
+                                lastRunNS = System.nanoTime();
+                                reroute = true;
+                                explanation = "one or more nodes has gone under the high or low watermark";
```

Comment:
Does it matter whether it went from above the high watermark to normal or above the low to normal? Could you get away with just storing a single Map? I figure that'd be simpler.